metrics.nim

```diff
@@ -542,6 +553,135 @@ template time*(summary: Summary, body: untyped) =
+                    buckets: bucketsSeq)
+    if labels.len == 0:
+      result.metrics[@labels] = newHistogramMetrics(name, labels, labels, bucketsSeq)
+    result.register(registry)
+
+template declareHistogram*(identifier: untyped,
+                        help: static string,
+                        labels: LabelsParam = @[],
+                        registry = defaultRegistry,
+                        buckets: openArray[float64] = defaultHistogramBuckets) {.dirty.} =
+  when defined(metrics):
+    var identifier = newHistogram(astToStr(identifier), help, labels, registry, buckets)
+  else:
+    type identifier = IgnoredCollector
+
+template declarePublicHistogram*(identifier: untyped,
+                               help: static string,
+                               labels: LabelsParam = @[],
+                               registry = defaultRegistry,
+                               buckets: openArray[float64] = defaultHistogramBuckets) {.dirty.} =
+  when defined(metrics):
+    var identifier* = newHistogram(astToStr(identifier), help, labels, registry, buckets)
+  else:
+    type identifier* = IgnoredCollector
+
+when defined(metrics):
+  proc histogram*(name: static string): Histogram =
+    var res {.global.} = newHistogram(name, "") # lifted line
+    return res
+else:
+  template histogram*(name: static string): untyped =
+    IgnoredCollector
+
+proc observeHistogram(histogram: Histogram, amount: int64|float64, labelValues: LabelsParam = @[]) =
+  when defined(metrics):
+    var timestamp = getTime().toMilliseconds()
+
+    let labelValuesCopy = validateHistogramLabelValues(histogram, labelValues)
+
+    atomicAdd(histogram.metrics[labelValuesCopy][0].value.addr, amount.float64) # _sum
+    atomicStore(cast[ptr int64](histogram.metrics[labelValuesCopy][0].timestamp.addr), timestamp.addr, ATOMIC_SEQ_CST)
+    atomicAdd(histogram.metrics[labelValuesCopy][1].value.addr, 1.float64) # _count
+    atomicStore(cast[ptr int64](histogram.metrics[labelValuesCopy][1].timestamp.addr), timestamp.addr, ATOMIC_SEQ_CST)
+    for i, bucket in histogram.buckets:
+      if amount.float64 <= bucket:
+        #- "le" probably stands for "less or equal"
+        #- the same observed value can increase multiple buckets, because this is
+        #  a cumulative histogram
+        atomicAdd(histogram.metrics[labelValuesCopy][i + 3].value.addr, 1.float64) # _bucket{le="<bucket value>"}
+        atomicStore(cast[ptr int64](histogram.metrics[labelValuesCopy][i + 3].timestamp.addr), timestamp.addr, ATOMIC_SEQ_CST)
+
+# the "type IgnoredCollector" case is covered by Summary.observe()
+template observe*(histogram: Histogram, amount: int64|float64 = 1, labelValues: LabelsParam = @[]) =
+  when defined(metrics):
+    {.gcsafe.}: observeHistogram(histogram, amount, labelValues)
+
+# in seconds
+# the "type IgnoredCollector" case is covered by Gauge.time()
+template time*(histogram: Histogram, labelValues: LabelsParam, body: untyped) =
```

Comment:
Isn't this definition essentially duplicated now for multiple metric types? You can define a single generic template (i.e. `metric: Histogram|Gauge|...`)

Reply:
In part. Gauge uses `set()` while Summary and Histogram use `observe()`.

Reply:
Similar templates combined.